src/ItemList.js

```diff
@@ -1,25 +1,88 @@
 import React, { useState, useEffect } from 'react';
 import Item from './Item';
 import { useHistory } from 'react-router-dom';
+import filter from './lib/filter';
+import { differenceInDays, fromUnixTime } from 'date-fns';
 
 function ItemList(props) {
   const [query, setQuery] = useState('');
-  const [queryArray, setQueryArray] = useState([]);
+  const [queryObj, setQueryObj] = useState({
+    week: [],
+    month: [],
+    longer: [],
+    inactive: [],
+  });
   let history = useHistory();
   const redirect = () => {
     history.push('/additems');
   };
 
   useEffect(() => {
-    // [filter-list] 2. Comparison function to filter shopping list and create a search results array
-    // [filter-list] (Remove accidental space character from query, make lowercase, and prevent regex errors)
-    const resultsArray = props.list.filter((itemObj) => {
-      return itemObj['itemName']
-        .toLowerCase()
-        .replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
-        .includes(query.toLowerCase().trim());
+    const resultsObj = {
+      week: [],
+      month: [],
+      longer: [],
+      inactive: [],
+    };
+
+    // Separate list items into categories based on how soon they should be bought
+    props.list.forEach((item) => {
+      if (
+        item.purchaseDates.length < 1 ||
+        !(item.purchaseEstimates?.length > 0)
+      ) {
+        resultsObj['inactive'].push(item);
+      } else {
+        const lastPurchase = fromUnixTime(
+          item.purchaseDates[item.purchaseDates.length - 1].seconds,
+        );
+        const lastInterval = differenceInDays(Date.now(), lastPurchase);
+        const lastEstimate =
+          item.purchaseEstimates[item.purchaseEstimates.length - 1];
+        const daysRemaining = lastEstimate - lastInterval;
+        item.daysRemaining = daysRemaining;
+
+        if (lastInterval >= 2 * lastEstimate) {
+          resultsObj['inactive'].push(item);
+        } else if (daysRemaining <= 7) {
+          resultsObj['week'].push(item);
+        } else if (daysRemaining <= 30) {
+          resultsObj['month'].push(item);
+        } else {
+          resultsObj['longer'].push(item);
+        }
+      }
```

Comment:
Minor performance question - does the resultsObj need to be recalculated if the query changes? Seems like this should only be recalculated if props.list changes.

Reply:
I think the reason we're recalculating it is to update the lists that are rendered on the page as the user updates their query. The default value of `resultsObj` is the full grocery list, and it is displayed on the page, and as the user searches, the value is filtered to match their query. Does that make sense?

Reply:
Yup that makes sense!

From the looks of this hook. You're always setting the default value of resultsObj after every change to query. Even though `props.list` hasn't changed (on line 29). 

Adding some console logs to the code, you can see that every time we add to a query, we loop through each item in the prop.list. 

<img width="411" alt="Screen Shot 2021-05-15 at 6 56 20 am" src="https://user-images.githubusercontent.com/3861088/118363735-ad062300-b54a-11eb-9817-76f50e8aaf12.png">

_Code for above screenshot_

```javascript
  useEffect(() => {
    const resultsObj = {
      week: [],
      month: [],
      longer: [],
      inactive: [],
    };

    console.log("Query is", query);
    console.log("Props list", props.list.length);
    // Separate list items into categories based on how soon they should be bought
    props.list.forEach((item) => {
      console.log("Looping through each item in props.list")
...
```

It seems like you should set the default value of resultsObj outside of this particular useEffect hook.

Reply:
Oh, I see what you mean. Thanks for the super helpful explanation! It makes sense that this wouldn't scale well for a bigger app, and the last time I ran lighthouse, our performance score wasn't amazing. I'll see if I can adjust this! Thanks, @jonathanyeong ! 😄

Reply:
@jonathanyeong @cafloyd I think I fixed it. 

I introduced another state variable `list` which stores the full list from props as an object with categories. This way, we don't have to sort items into categories every time the search query updates. Here are the console logs to confirm: https://www.loom.com/share/ccfc14ee754b46c39c98138d0ae61894

Then, I moved the filtering logic from `useEffect` to `changeHandler`. I tried to create another useEffect with `query` as a dependency because it seemed like a nicer way to do it. But it wouldn't reset the list when the search field was emptied and I'm now sure why. If there is a more efficient solution, please let me know.

Reply:
@dariatsvetkova nice, great work! I think we're now ready to merge?

Reply:
LGTM! Nice work @dariatsvetkova 🎉